src/transport/SessionManager.cpp

```diff
@@ -773,29 +779,60 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & packetHeade
         return;
     }
 
-    System::PacketBufferHandle msgCopy;
+    // Extract MIC from the end of the message.
+    uint8_t * data     = msg->Start();
+    uint16_t len       = msg->DataLength();
+    uint16_t footerLen = packetHeader.MICTagLength();
+    VerifyOrReturn(footerLen <= len); // err = CHIP_ERROR_INVALID_MESSAGE_LENGTH;
+
+    uint16_t taglen = 0;
+    MessageAuthenticationCode mac;
+    ReturnOnFailure(mac.Decode(packetHeader, &data[len - footerLen], footerLen, &taglen));
+    VerifyOrReturn(taglen == footerLen); // err = CHIP_ERROR_INTERNAL;
+
     bool decrypted = false;
     while (!decrypted && iter->Next(groupContext))
     {
+        CryptoContext context(groupContext.keyContext);
+        msgCopy = msg.CloneData();
+
+        if (packetHeader.HasPrivacyFlag())
+        {
+
+            // Perform privacy deobfuscation, if applicable.
+            uint8_t * privacyHeader = msgCopy->Start() + Header::kPrivacyHeaderOffset;
+            size_t privacyLength    = packetHeader.PrivacyHeaderLength();
+            ReturnOnFailure(context.PrivacyDecrypt(privacyHeader, privacyLength, privacyHeader, packetHeader, mac));
+        }
+
+        ReturnOnFailure(packetHeaderCopy.DecodeAndConsume(msgCopy));
```

Comment:
We need to ensure that older devices running the v1.0 branch don't have unexpected behaviors if a message that has the privacy flag is sent to them.

From what I understand of the existing code, it looks like they call this function today fairly early on in message processing in `SessionManager::OnMessageReceived`, which means they'll have bogus values for the source/dest/message counter fields in the message right?

Ultimately, it seems like that won't be a problem however, since by the time they get to the old logic in this function, it would have failed the call to `SecureMessageCodec::Decrypt`?

All in all, it would be really useful if you could run a test against the 1.0 branch to ensure this actually pans out, and that devices don't just crash.

Reply:
Yes, nodes running code before this patch will attempt to process garbage header fields.  The best coverage for these scenarios would be fuzzing tests.  I would like this patch to be merged into 1.0 as it is spec mandatory behavior to process privacy headers properly.

Reply:
Added `CHIP_CONFIG_PRIVACY_ACCEPT_NONSPEC_SVE2` as an option to handle the rare set of erroneous nodes that send with P=1 and no privacy prior to [#22425](https://github.com/project-chip/connectedhomeip/pull/22425/files#diff-76f701cb7a091ad07f50f468176175f53297e3efe37681a4f4205ecce234c93fL169).  Note the cost of enabling that is two rounds of decryption per key both with and without privacy. This PR includes extensive positive and negative unit tests for receiving group messages.

The `packetHeaderCopy` and `msgCopy` fields are designed to handle non-private fields such as message flags and session id and return remaining privacy decrypted fields after authentication.